preupg/creator/ui_helper.py

```diff
@@ -88,13 +88,13 @@ def _get_dict_value(self, name):
             return None
 
     def get_check_script(self):
-        return self._get_dict_value('check_script')
+        return 'check'
 
     def get_solution_file(self):
-        return self._get_dict_value('solution')
+        return 'solution.txt'
```

Comment:
Frankly I don't think it's a good practice to have literals inside a module just for the sake of it.  There's nothing wrong with having literals everywhere, as long as they are intrinsic to the object we're dealing with, IOW, they are never a subject to change.

In fact I'd say it's better to stick to literals; moving them to `settings` just forces reader to go there just to discover what it is.  `preupg` code is full of these detours and it's really annoying for me.

(Having a proper class loading and representing the object and dealing with all its external aspects would be the right thing, but that's not the case now AFAIK; all we're doing is tucking away literals).

`</rant>`

Reply:
I agree with @bocekm to have all things defined only once in one place. In the past, many issues have been delivered because of *definition* of the very same thing on more places.

Reply:
Well, if you can make typo in `return 'solution.txt'`, you can also make typo in `return settings.solution_txt`, right?

Anyway, sorry for the rant, none of this has really anything to do with this PR.  So do whatever needs to be done to finish it.

Reply:
But in one case it will crash because of variable that doesn't exist and in second case it would provide unexpected behaviour (especially in this project). In addition, in case we will want to change it in future to something else ("solution"?) doesn't matter whether yes or not now, I can imagine it will not be replaced on all places correctly. I saw similar incomplete changes in last years.

Anyway, discussion about something is always ok in case we will not spend on it hours :-) I am not saying that the replacement is the best solution. Just saying the benefit it has, which is better for me than save seconds for searching of values in the `settings`.

Reply:
If you want to account for future changes, then I have bad news for you: *you can't*¹.

In real life, when a change to API is made (I'm referring to module format as "API"), it's hardly ever that trivial: changes can remove/introduce mechanisms, alter behaviors and change meanings. (That is basicaly the point of changes, right? Nobody changes API just because they don't like a filename.)

Just try to imagine stretching the same concept into a real-world scenario³ of an API change; you will quickly realize that it just can't be done; you would end up with five times the amount of code (= five times the surface for bugs to live on) and in the end it would break anyway.

The point is, APIs **must be always treated by consumer as immutable**. The way we use of settings.py here is pretending otherwise, but that is a naive and dangerously wrong approach and should be avoided⁴.

Yes, you can imagine trivial cases such as we want to just rename a file or two², but that does not justify anything; that is just "what if (insert hypothetical scenario made up to justify status quo so that we don't have to admit how wrong our codebase is)".

Anyway, I'm willing to accept is as an exception if @bocekm's point was to keep the codebase consistent. (Although I would vote that the `settings` concept is wrong enough to justify inconsistency.)

---

¹ You actually can, by introducing API versioning.

² Note that unless you use black magic, such case is absolutely *trivial* to do in the codebase; simple `grep oldname -r` and maybe `find -name oldname` will solve 99% of cases.

³ ...and people actually never do it.  People write `foo >/dev/null`, not `foo >"$(load_name_of_dev_null)"`.  People don't account for API to magically change, because it's not *possible* to account for.

⁴ What could be done (and would be probably a good idea) would be to store all module facts and behaviors in a class, but that's entirely different question.

Reply:
@AloisMahdal, I have to disagree with some of your points. The fact that we shall conserve the API in an immutable state from the API user perspective is completely unrelated to the practice of using symbols for constants/literals occurring on many places and bearing the same meaning. From my perspective, it's more about keeping the code maintainable. If the symbol is named reasonably, you wouldn't even need to look for the value. The same applies to functions. You could argue in the same tone that it's hard for you to read the code because you need to jump there and back to read the content of those small functions. Then what you'd be looking for is actually spaghetti code. Instead, let the function do a single task and name the function based on this task.

I know there are cases of symbols in the _settings.py_ which are used just once, e.g. most of the messages printed to the user. That is in my eyes bad practice, which does not serve any reasonable purpose and such symbols are candidates to be placed directly to where they are used.

Anyway, here's my main point, why I'd like to use symbols instead of multiple direct placement of constants/literal across the codebase. Because then are able to tell whether the same literals have the same meaning. For example you could have the same literal for 1) a name of a directory, 2) a prefix for some complex string, 3) a name for a binary, etc.

I would rather lower your _find/grep_ estimate of 99% to maybe 80%. In those 20% you replace what you shouldn't and you break the functionality.

And, I agree with ⁴ :)